File: ios-app/Sources/RedCodeNetworking/ChatAPIEndpoint.swift
import Foundation

public enum ChatAPIEndpoint: Sendable {
    public static let chats = APIEndpoint(method: .get, path: "/chats")

    public static func deleteChat(roomID: String) -> APIEndpoint {
        APIEndpoint(method: .delete, path: "/chats/\(roomID)")
    }

    public static func messages(
        roomID: String,
        limit: Int = 50,
        beforeID: String? = nil,
        sinceID: String? = nil
    ) -> APIEndpoint {
        var queryItems = [
            URLQueryItem(name: "limit", value: String(limit)),
        ]
        if let beforeID, !beforeID.isEmpty {
            queryItems.append(URLQueryItem(name: "before_id", value: beforeID))
        }
        if let sinceID, !sinceID.isEmpty {
            queryItems.append(URLQueryItem(name: "since_id", value: sinceID))
        }
        return APIEndpoint(
            method: .get,
            path: "/rooms/\(roomID)/messages",
            queryItems: queryItems
        )
    }

    public static func sendMessage(roomID: String) -> APIEndpoint {
        APIEndpoint(method: .post, path: "/rooms/\(roomID)/messages")
    }

    public static func markMessagesRead(roomID: String) -> APIEndpoint {
        APIEndpoint(method: .post, path: "/rooms/\(roomID)/messages/read")
    }

    public static func deleteMessage(roomID: String, messageID: String) -> APIEndpoint {
        APIEndpoint(method: .delete, path: "/rooms/\(roomID)/messages/\(messageID)")
    }

    public static func pinMessage(roomID: String, messageID: String) -> APIEndpoint {
        APIEndpoint(method: .post, path: "/rooms/\(roomID)/messages/\(messageID)/pin")
    }

    public static func unpinMessage(roomID: String, messageID: String) -> APIEndpoint {
        APIEndpoint(method: .delete, path: "/rooms/\(roomID)/messages/\(messageID)/pin")
    }

    public static func messageReactions(roomID: String, messageID: String) -> APIEndpoint {
        APIEndpoint(method: .get, path: "/rooms/\(roomID)/messages/\(messageID)/reactions")
    }

    public static func addMessageReaction(roomID: String, messageID: String) -> APIEndpoint {
        APIEndpoint(method: .post, path: "/rooms/\(roomID)/messages/\(messageID)/reactions")
    }

    public static func removeMessageReaction(roomID: String, messageID: String, reactionKey: String) -> APIEndpoint {
        APIEndpoint(
            method: .delete,
            path: "/rooms/\(roomID)/messages/\(messageID)/reactions",
            queryItems: [
                URLQueryItem(name: "reaction_key", value: reactionKey),
            ]
        )
    }
}

File: android-app/app/src/main/java/com/redcode/im/androidapp/data/chat/ChatAPIEndpoint.kt
package com.redcode.im.androidapp.data.chat

import com.redcode.im.androidapp.network.APIEndpoint
import com.redcode.im.androidapp.network.HTTPMethod
import java.net.URLEncoder

object ChatAPIEndpoint {
    val chats = APIEndpoint(HTTPMethod.GET, "/chats")

    fun messages(roomId: String, limit: Int = 50, beforeId: String? = null, sinceId: String? = null): APIEndpoint =
        withQuery(
            path = "/rooms/$roomId/messages",
            values =
                mapOf(
                    "limit" to limit.coerceIn(1, 200).toString(),
                    "before_id" to beforeId,
                    "since_id" to sinceId,
                ),
        )

    fun sendMessage(roomId: String): APIEndpoint =
        APIEndpoint(HTTPMethod.POST, "/rooms/$roomId/messages")

    fun markMessagesRead(roomId: String): APIEndpoint =
        APIEndpoint(HTTPMethod.POST, "/rooms/$roomId/messages/read")

    private fun withQuery(path: String, values: Map<String, String?>): APIEndpoint {
        val query =
            values.entries
                .filter { (_, value) -> !value.isNullOrBlank() }
                .joinToString("&") { (key, value) ->
                    "${encode(key)}=${encode(value.orEmpty())}"
                }
        return APIEndpoint(HTTPMethod.GET, if (query.isBlank()) path else "$path?$query")
    }

    private fun encode(value: String): String =
        URLEncoder.encode(value, Charsets.UTF_8.name())
}
